---
title: "Exercise 06"
author: "Julia Frank"
date: "2025-03-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  library(motifmatchr) 
  library(Biostrings) 
  library(MotifDb) # database of motifs, not uptodate anymore
  library(TFBSTools) 
  library(universalmotif) 
  library(ensembldb)  
  library(epiwraps) 
  library(rtracklayer)
  library(BSgenome.Hsapiens.UCSC.hg38)
})
```

## Download the Data

**Choose a transcription factor, e.g. CREB1, REST, GATA5, EGR1, GCR (or any of your choice that has a motif and available ChIPseq data)**

I chose REST in Homo sapiens GM12878 (Lymphoblastoid cell lines (LCLs) generated by Epstein–Barr virus).

**Download the (e.g. Mouse) peaks for that factor (in whatever cell type)**

I used the latest conservative IDR-thresholded peaks (2021).

```{r}
options(timeout=3600) # Increase the download timeout
dir.create("raw_data") # Create a directory for the data
download.file("https://www.encodeproject.org/files/ENCFF428QEI/@@download/ENCFF428QEI.bed.gz", "raw_data/REST.bed.gz", mode="wb")
peaks_REST <- rtracklayer::import("raw_data/REST.bed.gz", format="narrowPeak")
```

**Identify the instances of the factor's motif**

Getting the desired motif

```{r}
motifs <- MotifDb::query(MotifDb, c("REST","Hsapiens"))
names(motifs)
# select newest HOCOMOCO motif
motif <- motifs[["Hsapiens-HOCOMOCOv11-core-A-REST_HUMAN.H11MO.0.A"]]
motif
view_motifs(motifs[1:2])
motif2 <- convert_motifs(motif, class="TFBSTools-PWMatrix")
```

Get the genome sequence

```{r}
genome_seqs <- getSeq(BSgenome.Hsapiens.UCSC.hg38)
```

## Answer the following questions:

-   **Of all the peaks, what proportion contains a motif for the factor?**

    **Expected form of an answer: of the XX peaks, XX (XX%) contain a motif**

```{r}
# Only look if peaks contain a motif 
moi <- motifmatchr::matchMotifs(motif2, subject=peaks_REST, genome=genome_seqs,
                                out="positions") 
moi <- moi[[1]] 

overlap <- sum(overlapsAny(peaks_REST, moi))
overlap_percent <- round(overlap/length(peaks_REST)*100, 2)
cat("of the ", length(peaks_REST), "peaks, ", overlap, "(", overlap_percent, "%) contain a motif")
```

-   **Of all instances of that motif in the genome (or in one chromosome), what proportion is bound by the factor (i.e. has a peak)?**

    **Expected form of an answer: of the XX motif instances, XX (XX%) overlap a peak**

```{r}
# Search for all motif instances across the whole genome
motif_across_genome <- matchMotifs(motif2, subject=genome_seqs, out="positions")[[1]]
names(motif_across_genome) <- names(genome_seqs)
motif_across_genome <- as(motif_across_genome, "GRanges")

overlap <- sum(overlapsAny(motif_across_genome, peaks_REST))
overlap_percent <- round(overlap/length(motif_across_genome)*100, 2)
cat("of the ", length(motif_across_genome), "motif instances,", overlap, "(", overlap_percent, "%) overlap a peak")
```
